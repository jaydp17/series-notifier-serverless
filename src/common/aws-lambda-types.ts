/**
 * Contains types that are related to AWS Lambda
 */

export type LambdaEvent = {
  httpMethod: string;
  body: string;
  queryStringParameters: {} | null;
};

/**
 * Callback type that should be used when the lambda is gonna be invoked with an HTTP req
 */
export type LambdaHttpCallback = (error: Error | null | undefined, response?: { statusCode: number, body?: string }) => void;

/**
 * Non-HTTP lambda callback type
 */
export type LambdaCallback = (error: Error | null | undefined, response?: {}) => void;
