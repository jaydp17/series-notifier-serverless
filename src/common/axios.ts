/**
 * Gives a common axios instance across the entire app
 */

import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import Logger from '../common/logger';

const timeLogger = Logger('TimeLogger');

function requestInterceptor(config: AxiosRequestConfig) {
  return { ...config, startTime: Date.now() };
}

function responseInterceptor(response: AxiosResponse) {
  const { startTime } = response.config as any; // tslint:disable-line:no-any
  const endTime = Date.now();
  timeLogger.log(`[TimeLogger]:: ${response.config.url} -> ${endTime - startTime}ms`);
  return response;
}

axios.interceptors.request.use(requestInterceptor);
axios.interceptors.response.use(responseInterceptor);

export default axios;
