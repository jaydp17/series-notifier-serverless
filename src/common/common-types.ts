/**
 * This file contains types that are consumed by more than 1 function
 */

export type Callback = (error: Error | null | undefined, data?: any) => void; // tslint:disable-line:no-any
