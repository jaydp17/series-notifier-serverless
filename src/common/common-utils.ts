/**
 * Contains utility functions
 */

import { inspect } from 'util';

/**
 * Pretty prints an object
 */
export function prettyPrint(obj: any): void { // tslint:disable-line:no-any
  console.log(inspect(obj, false, 20, true)); // tslint:disable-line:no-console
}
