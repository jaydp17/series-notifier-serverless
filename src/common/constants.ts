/**
 * Contains constants used accross the app
 */
export const GRAPH_API_URL = 'https://graph.facebook.com/v2.6';
