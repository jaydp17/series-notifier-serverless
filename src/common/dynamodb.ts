/**
 * Creates a dynamodb instance
 */

import { DynamoDB } from 'aws-sdk';
import { env } from './environment';

let dynamoDbOptions = {};
if (env === 'development' && process.env.IS_LOCAL) {
  dynamoDbOptions = { endpoint: 'http://localhost:8000' }; //tslint:disable-line
}
export const dynamodb = new DynamoDB(dynamoDbOptions);
export const dynamoDocClient = new DynamoDB.DocumentClient(dynamoDbOptions);

export default {
  get: params => dynamoDocClient.get(params).promise(),
  put: params => dynamoDocClient.put(params).promise(),
  query: params => dynamoDocClient.query(params).promise(),
  delete: params => dynamoDocClient.delete(params).promise(),
};
