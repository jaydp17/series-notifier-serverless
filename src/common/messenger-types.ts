interface Entry {
  id: string;
  time: number;
}

interface Messsaging {
  sender: { id: string; };
  recipient: { id: string; };
  timestamp: number;
}

export type TextMessage = {
  object: string;
  entry: TextMessageEntry[];
};

interface TextMessageEntry extends Entry {
  messaging: TextMessageMessaging[];
};

interface TextMessageMessaging extends Messsaging {
  message: {
    text: string;
    is_echo: boolean;
  }
};

export type AnyFacebookMessage = TextMessage;