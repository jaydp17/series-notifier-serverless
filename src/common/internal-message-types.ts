/**
 * Contains types related to messages transfered between lambas
 */

// tslint:disable:no-reserved-keywords

import * as MessengerTypes from '../common/messenger-types';
import { platformNames } from './constants';

/**
 * Represents the platforms from which a message can come & go
 */
export type Platform = 'FBMessenger' | 'AmazonAlexa' | 'GoogleHome';

export function isPlatform(platform: string): platform is Platform {
  return Object.values(<{ [key: string]: string }>platformNames).includes(platform);
}

/**
 * The processing function requires an action with an action type
 * This is the action type for it
 */
export enum ActionTypes {
  Search,
  Subscribe,
  UnSubscribe,
  ShowTrending,
  MyShows,
  NextEpisodeDate,
}

/**
 * The metaData object that's passed around across functions
 */
export interface IMetaData {
  fbMessenger?: MessengerTypes.AnyMessagingObject;
}

/**
 * The processing function requires an action to work on
 * It's similar to what Action/ActionTypes do in redux
 */
interface IAction {
  type: ActionTypes;
  platform: Platform;
  metaData: IMetaData;
}

/**
 * An action that represents a search query
 */
export interface ISearchAction extends IAction {
  type: ActionTypes.Search;
  text: string;
}

/**
 * An action that represents that the user is asking for the trending shows
 */
export interface IShowTrendingAction extends IAction {
  type: ActionTypes.ShowTrending;
}

/**
 * An action that represents that the user is asking to subscribe a show
 */
export interface ISubscribeAction extends IAction {
  type: ActionTypes.Subscribe;
  imdbId: string;
  tvdbId: number;
  title: string;
}

/**
 * An Action that represents that the user is trying to unSubscribe a show
 */
export interface IUnSubscribeAction extends IAction {
  type: ActionTypes.UnSubscribe;
  imdbId: string;
  tvdbId: number;
  title: string;
}

/**
 * An action that represents user asking for his subscribed shows
 */
export interface IMyShowsAction extends IAction {
  type: ActionTypes.MyShows;
}

export interface INextEpisodeDateAction extends IAction {
  type: ActionTypes.NextEpisodeDate;
  imdbId: string;
  tvdbId: number;
  title: string;
}

/**
 * Generic Action Type
 */
export type AnyAction =
  | ISearchAction
  | IShowTrendingAction
  | ISubscribeAction
  | IUnSubscribeAction
  | IMyShowsAction
  | INextEpisodeDateAction;

/**
 * Message that is sent to the ProcessQuery function
 * This is a cross platform way of telling that function to process something
 */
export interface IMessage {
  text: string;
  platform: Platform;
  metaData: {};
}

/**
 * Enum to classify the kind of reply
 */
export enum ReplyKind {
  Text,
  SearchResults,
  TrendingShows,
  SubscribeResult,
  UnSubscribeResult,
  MyShows,
  NextEpisodeDate,
  EpisodeNotification,
}

/**
 * The Formatter functions (i.e. the functions that reply back the the repective platforms)
 * get a reply object
 */
export interface IReply {
  kind: ReplyKind;
  metaData: IMetaData;
}

/**
 * A text message reply
 */
export interface ITextReply extends IReply {
  kind: ReplyKind.Text;
  text: string;
}

/**
 * A search results reply
 */
export interface ISearchResultsReply extends IReply {
  kind: ReplyKind.SearchResults;
  shows: ITvShow[];
}

/**
 * Reply to a subscribe show request
 */
export interface ISubscribeReply extends IReply {
  kind: ReplyKind.SubscribeResult;
  success: boolean;
  imdbId: string;
  title: string;
}

/**
 * Reply to an un-subscribe request
 */
export interface IUnSubscribeReply extends IReply {
  kind: ReplyKind.UnSubscribeResult;
  success: boolean;
  imdbId: string;
  title: string;
}

/**
 * A reply for Show Trending
 */
export interface ITrendingShowsReply extends IReply {
  kind: ReplyKind.TrendingShows;
  shows: ITvShow[];
}

/**
 * A reply for user's subscribed shows
 */
export interface IMyShowsReply extends IReply {
  kind: ReplyKind.MyShows;
  shows: ITvShow[];
}

/**
 * A reply for Next episode's date
 */
export interface INextEpisodeDateReply extends IReply {
  kind: ReplyKind.NextEpisodeDate;
  episode: ITvEpisode | null;
  error: { message: string } | null;
  show: {
    imdbId: string;
    title: string;
  };
}

/**
 * A reply for when the notification is sent to the user
 * when one of the episodes from the series he has subscribed goes live
 */
export interface IEpisodeNotificationReply extends IReply {
  kind: ReplyKind.EpisodeNotification;
  episode: ITvEpisode;
  show: {
    imdbId: string;
  };
}

/**
 * Generic reply type
 */
export type AnyReplyKind =
  | ITextReply
  | ISearchResultsReply
  | ITrendingShowsReply
  | ISubscribeReply
  | IUnSubscribeReply
  | IMyShowsReply
  | INextEpisodeDateReply
  | IEpisodeNotificationReply;

export interface ITvShowMetaData {
  title: string;
  year: number;
  tvdbId: number;
  imdbId: string;
  overview: string;
  genres: string[];
  backDropUrl?: string | null;
}

/**
 * Internal representation of a TV Show
 */
export interface ITvShow extends ITvShowMetaData {
  isSubscribed: boolean;
}

/**
 * Internal representation of an Episode
 */
export interface ITvEpisode {
  seasonNumber: number;
  epNumber: number;
  title?: string;
  tvdbId: number;
  imdbId: string;
  overview: string;
  rating?: number;
  firstAired?: number | null; // epoc time
  runtime?: number;
}
