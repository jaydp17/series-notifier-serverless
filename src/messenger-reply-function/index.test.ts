/**
 * Tests the reply function entry file
 */

jest.mock('../common/messenger.api');
jest.mock('./messenger.formatter');

import * as dummyCommonData from '../../test/test-data/common.data';
// types
import * as InternalTypes from '../common/internal-message-types';
import * as MessengerTypes from '../common/messenger-types';
import * as MessengerAPI from '../common/messenger.api';
import * as MessengerReply from './index';
import { generateGenericTemplate } from './messenger.formatter';

describe('Messenger Reply Function', () => {
  const senderId = '2342388586';
  const metaData: InternalTypes.IMetaData = {
    fbMessenger: <MessengerTypes.AnyMessagingObject>{
      sender: { id: senderId },
    },
  };

  beforeEach(() => {
    // clear mocks
    (<jest.Mock<{}>>MessengerAPI.sendMessage).mockClear();
    (<jest.Mock<{}>>generateGenericTemplate).mockClear();
  });

  it('checks for metaData [no metaData]', async () => {
    const reply = <InternalTypes.ITextReply>{
      kind: InternalTypes.ReplyKind.Text,
      text: 'hello',
    };
    const callback = jest.fn();

    // test
    await MessengerReply.handler(reply, {}, callback);
    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith(new Error('no metaData'));
  });

  it('checks for metaData [no fbMessenger metaData]', async () => {
    const reply = <InternalTypes.ITextReply>{
      kind: InternalTypes.ReplyKind.Text,
      metaData: { fbMessenger: {} },
      text: 'hello',
    };
    const callback = jest.fn();

    // test
    await MessengerReply.handler(reply, {}, callback);
    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith(new Error('No fbMessenger metaData'));
  });

  it('checks for senderId', async () => {
    const reply = <InternalTypes.ITextReply>{
      kind: InternalTypes.ReplyKind.Text,
      metaData: {
        fbMessenger: {
          sender: {},
        },
      },
      text: 'hello',
    };
    const callback = jest.fn();

    // test
    await MessengerReply.handler(reply, {}, callback);
    expect(callback).toHaveBeenCalledTimes(1);
    const error = (<jest.Mock<{}>>callback).mock.calls[0][0];
    expect(error).toBeInstanceOf(Error);
    expect(error.message).toMatch(/No senderId in metaData:/);
  });

  it('sends subscription result', async () => {
    const title = 'The Flash';
    const reply = <InternalTypes.ISubscribeReply>{
      kind: InternalTypes.ReplyKind.SubscribeResult,
      metaData,
      title,
    };
    const callback = jest.fn();

    // test
    await MessengerReply.handler(reply, {}, callback);
    expect(MessengerAPI.sendMessage).toHaveBeenCalledTimes(1);
    expect((<jest.Mock<{}>>MessengerAPI.sendMessage).mock.calls[0]).toMatchSnapshot();
    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith(null, { status: true });
  });

  it('sends un-subscription result', async () => {
    const title = 'The Flash';
    const reply = <InternalTypes.IUnSubscribeReply>{
      kind: InternalTypes.ReplyKind.UnSubscribeResult,
      metaData,
      title,
    };
    const callback = jest.fn();

    // test
    await MessengerReply.handler(reply, {}, callback);
    expect(MessengerAPI.sendMessage).toHaveBeenCalledTimes(1);
    expect((<jest.Mock<{}>>MessengerAPI.sendMessage).mock.calls[0]).toMatchSnapshot();
    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith(null, { status: true });
  });

  it('sends search results', async () => {
    const shows = [dummyCommonData.getTVShow()];
    const message = { payload: 'some-random-payload' };
    const reply = <InternalTypes.ISearchResultsReply>{
      kind: InternalTypes.ReplyKind.SearchResults,
      metaData,
      shows,
    };
    const callback = jest.fn();
    (<jest.Mock<{}>>generateGenericTemplate).mockReturnValueOnce(message);

    // test
    await MessengerReply.handler(reply, {}, callback);
    expect(generateGenericTemplate).toHaveBeenCalledTimes(1);
    expect(generateGenericTemplate).toHaveBeenCalledWith(shows);
    expect(MessengerAPI.sendMessage).toHaveBeenCalledTimes(1);
    expect(MessengerAPI.sendMessage).toHaveBeenCalledWith(senderId, message);
    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith(null, { status: true });
  });

  it('sends trending shows', async () => {
    const shows = [dummyCommonData.getTVShow()];
    const message = { payload: 'some-random-payload' };
    const reply = <InternalTypes.ITrendingShowsReply>{
      kind: InternalTypes.ReplyKind.TrendingShows,
      metaData,
      shows,
    };
    const callback = jest.fn();
    (<jest.Mock<{}>>generateGenericTemplate).mockReturnValueOnce(message);

    // test
    await MessengerReply.handler(reply, {}, callback);
    expect(generateGenericTemplate).toHaveBeenCalledTimes(1);
    expect(generateGenericTemplate).toHaveBeenCalledWith(shows);
    expect(MessengerAPI.sendMessage).toHaveBeenCalledTimes(1);
    expect(MessengerAPI.sendMessage).toHaveBeenCalledWith(senderId, message);
    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith(null, { status: true });
  });

  it('sends myShows results', async () => {
    const shows = [dummyCommonData.getTVShow()];
    for (let i = 0; i < 10; i += 1) {
      shows.push(dummyCommonData.getTVShow());
    }
    const message = { payload: 'some-random-payload' };
    const reply = <InternalTypes.IMyShowsReply>{
      kind: InternalTypes.ReplyKind.MyShows,
      metaData,
      shows,
    };
    const callback = jest.fn();
    (<jest.Mock<{}>>generateGenericTemplate).mockReturnValueOnce(message);

    // test
    await MessengerReply.handler(reply, {}, callback);
    expect(generateGenericTemplate).toHaveBeenCalledTimes(Math.ceil(shows.length / 10));
    expect(MessengerAPI.sendMessage).toHaveBeenCalledTimes(Math.ceil(shows.length / 10));
    expect(MessengerAPI.sendMessage).toHaveBeenCalledWith(senderId, message);
    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith(null, { status: true });
  });

  it('catches MessengerAPI.sendMessage error', async () => {
    const title = 'The Flash';
    const reply = <InternalTypes.ISubscribeReply>{
      kind: InternalTypes.ReplyKind.SubscribeResult,
      metaData,
      title,
    };
    const callback = jest.fn();
    const error = new Error('random error');
    (<jest.Mock<{}>>MessengerAPI.sendMessage).mockImplementationOnce(() => {
      throw error;
    });

    // test
    await MessengerReply.handler(reply, {}, callback);
    expect(callback).toHaveBeenCalledTimes(1);
    expect(callback).toHaveBeenCalledWith(error);
  });
});
