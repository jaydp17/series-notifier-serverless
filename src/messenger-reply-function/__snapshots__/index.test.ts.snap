// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messenger Reply Function sends subscription result 1`] = `
Array [
  "2342388586",
  Object {
    "text": "Subscribed to The Flash âœ…",
  },
]
`;

exports[`Messenger Reply Function sends un-subscription result 1`] = `
Array [
  "2342388586",
  Object {
    "text": "UnSubscribed from The Flash",
  },
]
`;
